great_expectations/core/util.py

```diff
@@ -782,148 +769,124 @@ def sniff_s3_compression(s3_url: S3Url) -> Union[str, None]:
+                ) from e
+            else:
+                raise e
 
-        spark_session = builder.getOrCreate()
-        # noinspection PyProtectedMember,PyUnresolvedReferences
-        if spark_session.sparkContext._jsc.sc().isStopped():
-            raise ValueError("SparkContext stopped unexpectedly.")
+        spark_session = _validate_spark_session_config(
+            spark_session=spark_session,
+            builder=builder,
+            spark_config=spark_config,
+        )
 
-    except AttributeError:
+    except AttributeError as e:
         logger.error(
             "Unable to load spark context; install optional spark dependency for support."
         )
-        spark_session = None
+        raise e
 
     return spark_session
 
 
-def spark_restart_required(
-    current_spark_config: List[Tuple[str, Any]], desired_spark_config: dict
-) -> bool:
-    """Determines whether or not Spark session should be restarted, based on supplied current and desired configuration.
+def _validate_spark_session_config(
+    spark_session: pyspark.SparkSession,
+    builder: _SparkSessionBuilder,
+    spark_config: dict,
+) -> pyspark.SparkSession:
+    if _spark_config_updatable(spark_session=spark_session):
+        if spark_config.get("spark.app.name"):
+            warnings.warn(
+                "Passing spark.app.name to spark_config has no effect in a Databricks environment.",
+                category=RuntimeWarning,
+            )
+    else:
+        # in a local pyspark-shell the context config cannot be updated
+        # unless you stop the Spark context and re-recreate it
+        retrieved_spark_config: pyspark.SparkConf = spark_session.sparkContext.getConf()
+        stopped = False
+        for key, value in spark_config.items():
+            if retrieved_spark_config.get(key) != value:
+                spark_session.stop()
+                stopped = True
+                break
+        if stopped:
+            spark_session = builder.getOrCreate()
+    return spark_session
 
-    Either new "App" name or configuration change necessitates Spark session restart.
 
-    Args:
-        current_spark_config: List of tuples containing Spark configuration string-valued key/property pairs.
-        desired_spark_config: List of tuples containing Spark configuration string-valued key/property pairs.
+def _spark_config_updatable(spark_session: pyspark.SparkSession) -> bool:
+    """
+    Tests whether we are able to update an existing Spark Session config.
 
-    Returns: Boolean flag indicating (if True) that Spark session restart is required.
+    Returns:
+        bool
     """
+    app_name: str
+    try:
+        app_name = spark_session.sparkContext.appName  # type: ignore[assignment]  # handled by try/except
+    except pyspark.PySparkNotImplementedError:
+        app_name = ""
+
+    updatable = "databricks" in app_name.lower()
+    return updatable
```

Comment:
It turns out we can't update a Spark/Databricks Connect session remotely, so only Databricks notebooks have updatable configs without stopping the Spark context.

Reply:
Couldn't a user have "databricks" in their app name even if it's a connect app?

Reply:
If it's a connect app, the `PySparkNotImplementedError` gets raised and the string is empty. I will add comments explaining that.